backport_packages/refactor_backport_packages.py

```diff
@@ -0,0 +1,750 @@
+
+        Adds provide context to usages of Python/BranchPython Operators in example dags.
+        Note that those changes  apply to example DAGs not to the operators/hooks erc.
+        We package the example DAGs together with the provider classes and they should serve as
+        examples independently on the version of Airflow it will be installed in.
+        Provide_context feature in Python operators was feature added 2.0.0 and we are still
+        using the "Core" operators from the Airflow version that the backport packages are installed
+        in - the "Core" operators do not have (for now) their own provider package.
+
+        The core operators are:
+
+            * Python
+            * BranchPython
+            * Bash
+            * Branch
+            * Dummy
+            * LatestOnly
+            * ShortCircuit
+            * PythonVirtualEnv
+
+
+        Example diff generated:
+
+        .. code-block:: diff
+
+            --- ./airflow/providers/amazon/aws/example_dags/example_google_api_to_s3_transfer_advanced.py
+            +++ ./airflow/providers/amazon/aws/example_dags/example_google_api_to_s3_transfer_advanced.py
+            @@ -105,7 +105,8 @@
+                         task_video_ids_to_s3.google_api_response_via_xcom,
+                         task_video_ids_to_s3.task_id
+                     ],
+            -        task_id='check_and_transform_video_ids'
+            +        task_id='check_and_transform_video_ids',
+            +        provide_context=True
+                 )
+
+        """
+        # noinspection PyUnusedLocal
+        def add_provide_context_to_python_operator(node: LN, capture: Capture, filename: Filename) -> None:
+            fn_args = capture['function_arguments'][0]
+            fn_args.append_child(Comma())
+
+            provide_context_arg = KeywordArg(Name('provide_context'), Name('True'))
+            provide_context_arg.prefix = fn_args.children[0].prefix
+            fn_args.append_child(provide_context_arg)
+
+        (
+            self.qry.
+            select_function("PythonOperator").
+            is_call().
+            is_filename(include=r"mlengine_operator_utils.py$").
+            modify(add_provide_context_to_python_operator)
+        )
+        (
+            self.qry.
+            select_function("BranchPythonOperator").
+            is_call().
+            is_filename(include=r"example_google_api_to_s3_transfer_advanced.py$").
+            modify(add_provide_context_to_python_operator)
+        )
+
+    def remove_super_init_call(self):
+        """
+        Removes super().__init__() call from Hooks. Example diff generated:
+
+        .. code-block:: diff
+
+            --- ./airflow/providers/apache/druid/hooks/druid.py
+            +++ ./airflow/providers/apache/druid/hooks/druid.py
+            @@ -49,7 +49,7 @@
+                         timeout=1,
+                         max_ingestion_time=None):
+
+            -        super().__init__()
+            +
```

Comment:
Why do we want to do this? Druid inherits basehook, surely we want to keep that still?

Reply:
I have a feeling that we discussed it several times already. Glad I am really patient :)

In airflow 1.10 none of the Hooks call super().init(). It was always broken in Airflow 1.10 - the BaseHook() has it's own __init__() which is wrongly implemented and requires source parameter to be passed::

```
    def __init__(self, source):
        pass
```

We fixed it in 2.0, but for the entire 1.10 line calling super().init() is not a good idea - and it basically does nothing even if you do. And it's bad because it does not initialize Logging Mixin (BaseHook derives from LoggingMixin). And it is the main reason why Hook logs are not working as they are supposed to sometimes:

```
class LoggingMixin(object):
    """
    Convenience super-class to have a logger configured with the class name
    """
    def __init__(self, context=None):
        self._set_context(context)
```

There are two Hooks in 1.10 that call super.__init__ :

```
       super(CloudSqlDatabaseHook, self).__init__(source=None)
       super(MongoHook, self).__init__(source='mongo')
```

Not that it helps with anything because init in BaseHook does nothing.

Reply:
BTW. I put it as a comment in the fixup I am sending shortly. I hope that will be the last time I am explaining it :)

Reply:
Ahhh thanks. I wasn't aware of this broken behaviour of BaseHook

I know I ask many things more than once -- this one I don't remember though.